File: src/variant_1/MyThreadLocal.java
package variant_1;

import java.util.HashMap;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/*
* Принцип данного подхода состоит в создании некой области памяти,
* куда потоки смогут сохранять свои данные. Общение потока и SharedMemory производится через
* класс MyThreadLocal, в котором реализованы функции put, get, remove.
* */
public class MyThreadLocal<T> {
    private final UUID uuid;

    MyThreadLocal() {
        uuid = UUID.randomUUID();
    }

    public T get() {
        return (T) SharedMemory.getValue(uuid);
    }

    public void remove() {
        SharedMemory.removeValue(uuid);
    }

    public void put(T value) {
        SharedMemory.putValue(uuid, value);
    }
}

/*
 * Класс, который реализует потокобезопасную запись, чтение, удаление "локальных" переменных потоков.
 * */
class SharedMemory {
    private static final ConcurrentHashMap<UUID, Object> storage = new ConcurrentHashMap();

    public static void putValue(UUID key, Object value) {
        storage.put(key, value);
    }

    public static Object getValue(UUID key) {
        return storage.get(key);
    }

    public static void removeValue(UUID key) {
        storage.remove(key);
    }
}

File: src/storage/MyThreadLocal.kt
package storage

import java.util.concurrent.atomic.AtomicInteger

/*
* Объекты класса предназначены для сохранения данных конкретного потока.
* */
class MyThreadLocal<T> {
    private var storage: MyThreadLocalStorage<T> = MyThreadLocalStorage()

    fun get(): T? {
        return storage.getValue(Thread.currentThread().id)
    }

    fun set(value: T) {
        println(Thread.currentThread().id)
        storage.setValue(Thread.currentThread().id, value)
    }

    fun remove() {
        storage.removeValue(Thread.currentThread().id)
    }

    /*
    * Хранилище, локальное для каждого потока.
    * */
    class MyThreadLocalStorage<T> {
        private var map: HashMap<Long, T> = hashMapOf()

        fun setValue(key: Long, value: T) {
            map[key] = value
        }

        fun removeValue(hash: Long) {
            map.remove(hash)
        }

        fun getValue(hash: Long): T? {
            return map[hash]
        }
    }
}

